src/lib.rs

```diff
@@ -260,29 +350,82 @@ impl<B: MysqlShim<W>, R: Read, W: Write> MysqlIntermediary<B, R, W> {
                     "peer terminated connection",
                 )
             })?;
-            let _handshake = commands::client_handshake(&handshake)
+            let handshake = commands::client_handshake(&handshake)
                 .map_err(|e| match e {
                     nom::Err::Incomplete(_) => io::Error::new(
                         io::ErrorKind::UnexpectedEof,
                         "client sent incomplete handshake",
                     ),
-                    nom::Err::Failure((input, nom_e_kind))
-                    | nom::Err::Error((input, nom_e_kind)) => {
-                        if let nom::error::ErrorKind::Eof = nom_e_kind {
+                    nom::Err::Failure(nom_error) | nom::Err::Error(nom_error) => {
+                        if let nom::error::ErrorKind::Eof = nom_error.code {
                             io::Error::new(
                                 io::ErrorKind::UnexpectedEof,
-                                format!("client did not complete handshake; got {:?}", input),
+                                format!(
+                                    "client did not complete handshake; got {:?}",
+                                    nom_error.input
+                                ),
                             )
                         } else {
                             io::Error::new(
                                 io::ErrorKind::InvalidData,
-                                format!("bad client handshake; got {:?} ({:?})", input, nom_e_kind),
+                                format!(
+                                    "bad client handshake; got {:?} ({:?})",
+                                    nom_error.input, nom_error.code
+                                ),
                             )
                         }
                     }
                 })?
                 .1;
+
+            let mut auth_response = handshake.auth_response.clone();
+            let auth_plugin_expect = self.shim.auth_plugin_for_username(&handshake.username);
+
+            // auth switch
+            if !auth_plugin_expect.is_empty()
+                && auth_response.is_empty()
+                && handshake.auth_plugin != auth_plugin_expect.as_bytes()
+            {
+                self.writer.write_all(&[0xfe])?;
+                self.writer.write_all(auth_plugin_expect.as_bytes())?;
+                self.writer.write_all(&[0x00])?;
+                self.writer.write_all(&scramble)?;
+                self.writer.write_all(&[0x00])?;
```

Comment:
If this is an error packet, let's re-use the existing mechanisms for writing an error packet rather than hard-coding the values here.

Reply:
If it's not an error packet, a comment explaining what's going on would be nice.